internal/proxy/connect_handler_test.go

```diff
@@ -42,203 +47,344 @@ func newConnHandlerTestDepsConfig(proxyEndpoint string) connHandlerTestDepsConfi
-	node := tools.NodeWithMemoryEngineNoHandlers()
-	defer func() { _ = node.Shutdown(context.Background()) }()
+func teardownConnHandleHTTPTestCase(c httpConnHandleTestCase) {
+	defer func() { _ = c.node.Shutdown(context.Background()) }()
+	c.server.Close()
+}
 
-	mux := http.NewServeMux()
-	mux.Handle("/proxy", &proxyHandler{})
-	server := httptest.NewServer(mux)
-	defer server.Close()
+type connHandleTestCase struct {
+	connectProxyHandler *ConnectHandler
+	protocol            string
+}
 
-	testDepsCfg := newConnHandlerTestDepsConfig(server.URL + "/proxy")
-	connHandler := testDepsCfg.connectProxyHandler.Handle(node)
+func (c connHandleTestCase) invokeHandle(ctx context.Context, httpTestCase httpConnHandleTestCase, grpcTestCase grpcConnHandleTestCase, currentCase connHandleTestCase) (reply centrifuge.ConnectReply, err error) {
+	var connHandler centrifuge.ConnectingHandler
 
-	ctx, cancel := context.WithCancel(context.Background())
-	cancel()
+	if currentCase.protocol == "http" {
+		connHandler = currentCase.connectProxyHandler.Handle(httpTestCase.node)
+		reply, err = connHandler(ctx, httpTestCase.cfg.connectEvent)
+	} else if currentCase.protocol == "grpc" {
+		connHandler = currentCase.connectProxyHandler.Handle(grpcTestCase.node)
+		reply, err = connHandler(ctx, grpcTestCase.cfg.connectEvent)
+	}
 
-	connReply, err := connHandler(ctx, testDepsCfg.connectEvent)
-	require.ErrorIs(t, err, centrifuge.DisconnectNormal)
-	require.Equal(t, connReply, centrifuge.ConnectReply{})
+	return reply, err
 }
 
-func TestHandleWithoutProxyServerStart(t *testing.T) {
-	node := tools.NodeWithMemoryEngineNoHandlers()
-	defer func() { _ = node.Shutdown(context.Background()) }()
+func newConnHandleTestCases(httpTestCase httpConnHandleTestCase, grpcTestCase grpcConnHandleTestCase) []connHandleTestCase {
+	return []connHandleTestCase{
+		{
+			connectProxyHandler: grpcTestCase.cfg.connectProxyHandler,
+			protocol:            "grpc",
+		},
+		{
+			connectProxyHandler: httpTestCase.cfg.connectProxyHandler,
+			protocol:            "http",
+		},
+	}
+}
+
+func TestHandleConnectWithEmptyReply(t *testing.T) {
+	grpcTestCase := newConnHandleGRPCTestCase(newProxyGRPCTestServer("", proxyGRPCTestServerOptions{}))
+	defer teardownConnHandleGRPCTestCase(grpcTestCase)
 
-	testDepsCfg := newConnHandlerTestDepsConfig("/proxy")
-	connHandler := testDepsCfg.connectProxyHandler.Handle(node)
-	connReply, err := connHandler(context.Background(), testDepsCfg.connectEvent)
-	require.ErrorIs(t, err, centrifuge.ErrorInternal)
-	require.Equal(t, connReply, centrifuge.ConnectReply{})
+	httpTestCase := newConnHandleHTTPTestCase("/proxy")
+	httpTestCase.mux.HandleFunc("/proxy", func(w http.ResponseWriter, req *http.Request) {
+		_, _ = w.Write([]byte(`{}`))
+	})
+	defer teardownConnHandleHTTPTestCase(httpTestCase)
+
+	cases := newConnHandleTestCases(httpTestCase, grpcTestCase)
+	for _, c := range cases {
+		reply, err := c.invokeHandle(context.Background(), httpTestCase, grpcTestCase, c)
```

Comment:
So the idea here was avoid passing `httpTestCase`, `grpcTestCase` and `c` to `c.invokeHandle`, i.e. have sth like this:

```
cases := newConnHandleTestCases(httpTestCase, grpcTestCase)
for _, c := range cases {
  reply, err := c.invokeHandle(context.Background())
}
```

Does this make sense?

Reply:
I also don't like that it is necessary to explicitly pass `httpTestCase` and `grpcTestCase`.
I'll think about whether it can be done in a simple way.

Reply:
Done.
Seems to look better.

Reply:
Great 👍